DESIGN: document  the behaviour and implementation of CRs Approved/Denied conditions (#3797)
jetstack/cert-manager: design/20190708.certificate-request-crd.md

##### RBAC

Approved and Denied conditions are set by requesting against the `/status`
endpoint of the CertificateRequest resource. This is a divergence of the
certificates.k8s.io CertificateSigningRequest resource that utilises `approval`,
as CustomResourceDefinitions may [only define `/status` or `/scale`
sub-resources](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#subresources).
Approvers must therefore have permissions to update the status sub-resource of
CertificateRequests. Approvers missing this permissions will have their request
rejected by the API server.

Setting the Approved or Denied conditions are restricted by the approver having
sufficient RBAC permissions. These permissions are based upon the request
itself- specifically the request's IssuerRef:

```yaml
apiGroups: ["cert-manager.io"]
resources: ["signers"]
verbs: ["approve"]
resourceNames:
 # namesapced signers
 - "<signer-resource-name>.<signer-group>/<signer-namespace>.<signer-name>"
 # cluster scoped signers
 - "<signer-resource-name>.<signer-group>/<signer-name>"
 # all signers of this resource name
 - "<signer-resource-name>.<signer-group>/*"

> ```

Reviewer: Might be worth changing "resource name" to "resource type" since it seems to be the official term for it (cf. [api-concepts](https://kubernetes.io/docs/reference/using-api/api-concepts/))

I also propose to clarify what each `resourceName` syntax does:

~~~suggestion
```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: Role
rules:
- apiGroups: ["cert-manager.io"]
  resources: ["signers"]
  verbs: ["approve"]
  resourceNames:
   # To allow a namespaced signer, use the following syntax:
   - "<signer-resource-type>.<signer-group>/<signer-namespace>.<signer-resource-name>"
   # To allow a cluster-scoped signer, use the following syntax:
   - "<signer-resource-type>.<signer-group>/<signer-resource-name>"
   # To allow a all the signers for a given group:
   - "<signer-resource-type>.<signer-group>/*"
```
~~~

An example ClusterRole that would grant the permissions to set the Approve and
Denied conditions of CertificateRequests that reference the cluster scoped
`myissuers` external issuer, in the group `my-example.io`, with the name `myapp`:

```
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: my-example-io-my-issuer-myapp-approver
rules:
  - apiGroups: ["cert-manager.io"]
    resources: ["signers"]
    verbs: ["approve"]

>     resourceNames: ["myissuers.my-example.io/myapp"]

Reviewer: what is "myapp"?

Author: Yes

Reviewer: `myapp` is an uncommon name for an issuer, isn't it? I propose to use a name that help the reader understand that this name corresponds to the issuer object itself:
```suggestion
    resourceNames: ["myissuers.my-example.io/issuer-1"]
```

> ```

Reviewer: Would be nice to use concrete names in the example to help the reader understand what this new RBAC syntax entails. I also propose to add a concrete example:

~~~suggestion

Let us imagine that we want to build an approver that uses OPA (Open Policy Agent). Our approver
takes the form of a Kubernetes controller that watches the CertificateRequest objects. For every
CertificateRequest, our approver asks the OPA engine whether the CertificateRequest should
be accepted or not:

- If OPA returns true, the approver adds the `Approved=True` condition to the CertificateRequest;
- If OPA returns false, the approver adds the `Denied=True` condition to the CertificateRequest.

By default, our approver does not have the permission to add the `Approved=True`
and `Denied=True` conditions. In order to allow it to add these conditions, we use a custom
RBAC syntax that cert-manager knows about. Let us imagine that our approver needs to approve
CertificateRequests that are targeting the internal ACME issuer as well as the external issuer for
[AWS PrivateCA](https://github.com/jniebuhr/aws-pca-issuer):

```
apiVersion: cert-manager.io/v1
kind: ClusterIssuer
metadata:
  name: letsencrypt-issuer
  namespace: some-namespace
spec:
  acme:
    email: jane.doe@gmail.com
    server: https://acme-staging-v02.api.letsencrypt.org/directory
    privateKeySecretRef:
      name: letsencrypt-issuer-account
    solvers:
      - http01:
          ingress:
            class: traefik
---
apiVersion: awspca.cert-manager.io
kind: AWSPCAClusterIssuer
metadata:
  name: awspca-issuer
  namespace: some-namespace
spec:
  arn: "pca-807911e9"
```

To make our approver work for all CertificateRequests across all namespaces,
we apply the following role to our cluster:

```
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: opa-approver-for-acme-and-aws-pca
rules:
  - apiGroups: ["cert-manager.io"]
    resources: ["signers"]
    verbs: ["approve"]
    resourceNames:
      - "clusterissuers.acme.cert-manager.io/letsencrypt-issuer"
      #  <-------------------------------->  <---------------->
      #    resource type + resource group        issuer name
      - "awspcaclusterissuers.awspca.cert-manager.io/awspca-issuer"
```

Note: the "resource type" [is](https://kubernetes.io/docs/reference/using-api/api-concepts/#standard-api-terminology)
the lower-case and plural version of the "kind" field.
~~~

Author: I think this would be a good addition to the website, where if someone wanted to write an approver.

> These permissions are enforced using a
> [SubjectAccessReview](https://github.com/kubernetes/kubernetes/blob/b11d0fbdd58394a62622787b38e98a620df82750/pkg/apis/authorization/types.go#L27)
> that is performed in the cert-manager Webhook component when a user attempts to
> set the Approve or Denied condition. The user of the approver is tested along
> with the IssuerRef of the CertificateRequest.

Reviewer: ```suggestion
These permissions are enforced using the
[SubjectAccessReview](https://github.com/kubernetes/kubernetes/blob/b11d0fbdd58394a62622787b38e98a620df82750/pkg/apis/authorization/types.go#L27). 

The cert-manager webhook looks at the SubjectAccessReview provided by the user (or, in the above
example, the approver) and checks that they have the correct RBAC as defined above.
```

Author: It is the webhook itself that actually performs the SAR request, rather than the user supplying it to the webhook.

If the approver does not have sufficient permissions defined above to set the
Approved or Denied conditions, the request will be rejected by the cert-manager
validating admission webhook.

> ###### Scope

Reviewer: Empty line like the above sections

```suggestion
###### Scope

```

The RBAC permissions *must* be granted at the cluster scope (i.e.
ClusterRoleBinding). This matches the same scoping behaviour as
[certificates.k8s.io

> CertificateSigningRequests](https://kubernetes.io/docs/reference/access-authn-authz/certificate-signing-requests/#authorization).

Reviewer: Why should it only be at the cluster scope?

Author: This is the same behaviour as the upstream CSR resource, which is cluster scoped. The namespace is encoded within the signer ref (exmaple.io/my-namespace.bar)

Namespaced signers are represented by a namespaced resource using the syntax of
`<signer-resource-name>.<signer-group>/<signer-namespace>.<signer-name>`

###### API Group
The apiGroup must *always* be `cert-manager.io`, as all CertificateRequests are
in the `cert-manager.io` group. Even though some signers may themselves not
belong to the `cert-manager.io` group, the approve operation does.

###### Resource
The resource must *always* be `signers`. This matches the same resource as
[required by certificates.k8s.io
CertificateSigningRequests](https://kubernetes.io/docs/reference/access-authn-authz/certificate-signing-requests/#authorization).

###### Verbs
The verb must *always* be `approve`. This verb grants the approver the
permissions to set *both* Approved and Denied conditions.

###### Resource Names
The resource name is made up of 3 parts: the signer resource name, signer group, and the

> signer name.

Reviewer: Could we use the official terminology "issuer" instead of "signer"? Signer is not a term we officially use in the cert-manager project.

Author: This is a deliberate decision, to again align with upstream. We should start migrating our terminology.

- A signer is an entity which signs
- An Issuer is a cert-manager specific implementation of a signer

A signer may or may not be a cert-manager Issuer.

Reviewer: I think it is worth adding this note to the top of the design doc.

> ###### API Group
> The apiGroup must *always* be `cert-manager.io`, as all CertificateRequests are
> in the `cert-manager.io` group. Even though some signers may themselves not
> belong to the `cert-manager.io` group, the approve operation does.
> 
> ###### Resource
> The resource must *always* be `signers`. This matches the same resource as
> [required by certificates.k8s.io
> CertificateSigningRequests](https://kubernetes.io/docs/reference/access-authn-authz/certificate-signing-requests/#authorization).
> 
> ###### Verbs
> The verb must *always* be `approve`. This verb grants the approver the
> permissions to set *both* Approved and Denied conditions.
> 
> ###### Resource Names
> The resource name is made up of 3 parts: the signer resource name, signer group, and the
> signer name.

Reviewer: It would be worth using the typography `apiGroups`, `resources`, `verbs` and `resourceNames` in these 4 sections to help the reader understand that these subsections refer to the above RBAC role.

I suggest using a table:

```suggestion
|      Field      |                                                      Description                                                       |
|-----------------|------------------------------------------------------------------------------------------------------------------------|
| `apiGroups`     | Must be `cert-manager.io`. This is required for the `approve` verb, even if an external issuers has a different group. |
| `resources`     | Must be `signers`. This is similar to [CertificateSigningRequests][csr].                                               |
| `verbs`         | Must be `signers`. Gives permission to set the `Approved` and `Denied` conditions.                                     |
| `resourceNames` | Uses a custom syntax described below.                                                                                  |

[csr]: https://kubernetes.io/docs/reference/access-authn-authz/certificate-signing-requests/#authorization

###### resourceNames custom syntax
```

Author: Thanks, I will update the names. Not keen on tables at it makes it more difficult to read in a text editor.

The signer resource name should be the same string of the resource name of the
signer.

> The signer group should be the same string of the group of signer.

Reviewer: ```suggestion
The signer group should be the same string as the group of the issuer.
```

The name may be either a wild card '*' denoting all signer names in all
namespaces, or instead a string which must match optionally a namespace,
followed by the signer name as it appears in the IssuerRef.

An example of signing all `myissuer` signers in all namespaces, and

> `clustermyissuers` with the name `myppp`, in the `my-example.io` group:

Reviewer: myppp?

```yaml
    resourceNames: ["myissuers.my-example.io/*", "clustermyissuers.my-example.io/myapp"]

> ```

Reviewer: ~~~suggestion
```yaml
    resourceNames:
      - myissuers.my-example.io/*             # Any MyIssuer object can approve.
      - clustermyissuers.my-example.io/myapp  # Only the ClusterMyIssuer object that is named "myapp" can approve.
```
~~~

An example of signing `myissuer` with the name `myapp` in the namesapces `foo`
and `bar`:

```yaml
    resourceNames: ["myissuers.my-example.io/foo.myapp", "myissuers.my-example.io/bar.myapp"]

> ```

Reviewer: Suggestion: the reader might find it easier to understand with some comments:

~~~suggestion
```yaml
resourceNames:
  # With the following, the user (i.e., the approver) can only approve the CertificateRequests
  # in the namespace "foo" and that are using the namespace-scoped MyIssuer named
  # "myapp":
  - myissuers.my-example.io/foo.myapp       
  # Similar to the above, but in the namespace "bar":
  - myissuers.my-example.io/bar.myapp
```
~~~

> ##### Webhook signer string building

Reviewer: signer → issuer

The validating admission webhook is responsible for making SubjectAccessReviews
to evaluate whether the approver has [sufficient permissions to add the Approved
or Denied conditions](#rbac). The webhook will attempt to first check whether
the use can approve all signers in all namespaced (wildcard), and if this fails,
whether it can approve for the exact signer (`.<signer-name>` or
`<signer-namespace>.<signer-name>`).

The webhook will keep a cache of the [Discovery
API](https://github.com/kubernetes/client-go/blob/f6ce18ae578c8cca64d14ab9687824d9e1305a67/discovery/discovery_client.go#L55)
which will be used to determine whether a referenced signer is namespaced or
not. If it is namespaced, the `<signer-namespace>` will be populated with the
namesapce that the CertificateRequest resides in. If the scope of the resource
cannot be determined, the request will be rejected.

##### cert-manager Approver

The cert-manager controller is deployed with an internal approver controller.

> This controller will attempt to Approve *all* CertificateRequests that are

Reviewer: Why do we capitalize "approve" here?

```suggestion
This controller will attempt to approve *all* CertificateRequests that are
```

> created, regardless of its contents or IssuerRef.

Reviewer: ```suggestion
created, regardless of the issuer they reference or their contents.
```

By default, this approver is given the permission to approve all internal

> signers, that is `Issuer` and `ClusterIssuer` `cert-manager.io` signers. This

Reviewer: signers → issuers

means its ClusterRole is given the `resourceNames`: `issuers.cert-manager.io/*`
and `clusterissuers.cert-manager.io/*`

External issuers may consider whether they wish to install further RBAC that
would allow the cert-manager approver controller to also approve
CertificateRequests that reference these external issuers. This would involve
creating a ClusterRole that permits approving that signer type, and binding that
role to the cert-manager-controller ServiceAccount.

The default cert-manager approver controller may be disabled by adding the
`--controllers=*,-certificaterequests-approver` argument to the
`cert-manager-controller` component. This allows for other approvers to make
decisions about CertificateRequests, without racing against the internal
approver controller.

### Controller Behaviour

> The philosophy for the `CertificateRequest` controllers are planned to be as

Reviewer: are planned → is planned (philosophy is singular)

This controller will attempt to Approve *all* CertificateRequests that are
created, regardless of its contents or IssuerRef.

> By default, this approver is given the permission to sign all internal signers,

Reviewer 2: Nit: 'to sign all' -> 'to approve all' ?

further processing will take place by the `CertificateRequest` controller on
this resource.

#### Issuing Controller

Since external issuers have been built before the addition of Approved and
Denied conditions, the issuing controller needs to be permissive. An external
issuer may not honour an Approved condition and will sign and set a
CertificateRequest as being Ready, before the request has been approved. The
issuing controller must mark issuance as being successful in this case. In
practice, this means that the issuing controller is never concerned with
Approved conditions.

External issuers that do not honour Denied conditions will sign
CertificateRequests, even if they have a Denied condition set. In this case, the
issuing controller will successfully complete the issuance of the Certificate.

> External Issuers and internal issuer that honour the Denied condition will never

Reviewer: "Issuer" is capitalized, "issuer" is not. Does it denote a subtle difference in semantics that I may not be able to see?

Other remark: I thought we had settled on the US spelling (but we did not either write it down or say that it is only for the cert-manager codebase...)

```suggestion
External issuers and internal issuer that honor the Denied condition will never
```

sign CertificateRequests with the Denied condition set, and thus never set the

> Ready condition. In this case, the issuing controller will consider this

Reviewer: By "issuing controller", do you refer to external issuers? or internal issuers? or both?

Author: Both

Reviewer: I realize that I am confusing "issuing controller" with [issuing_controller.go](https://github.com/jetstack/cert-manager/blob/master/pkg/controller/certificates/issuing/issuing_controller.go). Could we rephrase to remove the ambiguity?

Also, I am not too sure about what "in this case" refers to. Does it refer to the fact that the issuer is honoring?

```suggestion
Ready condition. We say that an issuer is "honoring" the Approval API when the issuer considers this
```

> Ready condition. In this case, the issuing controller will consider this
> CertificateRequest as failed, set the Issuing condition to False, and set the
> last failure time. This will cause the issuance to be retried at a later date (1

Reviewer: I know it has already been mentioned above, but it would be worth adding a note about the requirements for approvers that are honoring the Approval API: they must set `Ready=False` on CertificateRequests when they notice that the CR is either `Approved=True` or `Denied=True`

Suugestion:

```suggestion
Ready condition. In this case, the issuing controller will consider this
CertificateRequest as failed, and will set the condition `Issuing=False`
as well as setting the status field `lastFailureTime`. Note that the issuing
controller is not responsible for setting the `Ready=False` condition on
the CertificateRequests; that's the issuer's responsibility.

The Certificate is then retried later (1
```
